jni/src/test/java/com/horizen/sc2scnative/Sc2ScTest.java

```diff
@@ -0,0 +1,290 @@
+    }
+
+    private byte[] getMsgRoot() {
+        msgTree.finalizeTreeInPlace();
+        try (FieldElement root = msgTree.root()) {
+            return root.serializeFieldElement();
+        }
+    }
+
+    @Test
+    public void happyPath() {
+        Random r = new Random();
+        RandomProofData proofData = generateRandomProof(r, 42);
+
+        assertNotNull("Proof creation must be successful", proofData.proof);
+
+        assertEquals(psType, ProvingSystem.getProofProvingSystemType(proofData.proof));
+
+        assertTrue(Sc2Sc.verifyProof(
+                proofData.nextScTxCommitmentsRoot,
+                proofData.currentScTxCommitmentsRoot,
+                proofData.msgHash,
+                proofData.proof,
+                vkPath));
+    }
+
+    private byte[] changeBytes(byte[] data) {
+        data[0] = (byte) (data[0] + 1);
+        return data;
+    }
+
+    @Test
+    public void shouldNotVerifyInvalidInput() {
+        Random r = new Random();
+        RandomProofData proofData = generateRandomProof(r, 42);
+
+        assertFalse(
+                Sc2Sc.verifyProof(
+                        changeBytes(proofData.nextScTxCommitmentsRoot),
+                        proofData.currentScTxCommitmentsRoot,
+                        proofData.msgHash,
+                        proofData.proof,
+                        vkPath));
+
+        assertFalse(
+                Sc2Sc.verifyProof(
+                        proofData.nextScTxCommitmentsRoot,
+                        changeBytes(proofData.currentScTxCommitmentsRoot),
+                        proofData.msgHash,
+                        proofData.proof,
+                        vkPath));
+        assertFalse(
+                Sc2Sc.verifyProof(
+                        proofData.nextScTxCommitmentsRoot,
+                        proofData.currentScTxCommitmentsRoot,
+                        changeBytes(proofData.msgHash),
+                        proofData.proof,
+                        vkPath));
+    }
+
+    // Generate a RandomProofData with just a a certificate for epoch and epoch + 1.
+    private RandomProofData generateRandomProof(Random r, int epoch) {
+        // We'll create tx commitment for a epoch with one certificate and
+        // another one for next epoch with another certificate. From this
+        // commitments we'll extract the path and then create a circuit.
+        byte[] scId = generateFieldRandomBytes(r);
+        byte[] msgHash = appendRandomMsgHash(r);
+        byte[] msgRoot = getMsgRoot();
+
+        WithdrawalCertificate currWithdrawalCertificate = WithdrawalCertificate.getRandom(
+                r, 0, MIN_CUSTOM_FIELDS);
+        currWithdrawalCertificate.setScId(scId);
+        currWithdrawalCertificate.setEpochNumber(epoch);
+
+        currWithdrawalCertificate.setCustomField(MSG_ROOT_HASH_CUSTOM_FIELDS_POS, msgRoot);
+        byte[] currCertHash = currWithdrawalCertificate.getHashBytes();
+
+        addRandomScc(r, currentCt, scId);
```

Comment:
don't know if it relevant for the precision of the test, but it can't happen that in the same tree we have for the same scid a sidechain creation and a a certificate containing messages  (creation happens in the genesys block without any other operation)

Reply:
we should test also filling the tree with random data coming from other sidechains (remember the tree is a "snaposhot" of all sidechain related stuff happened in a timeframe, so it may contain for example also forward transfers, bwt form different sidehcian)

Reply:
On the rust side I've tested all negative cases and there is also a complete test that use more sidechains and both forward and backward transfers.